worldedit-core/src/main/java/com/sk89q/worldedit/internal/util/RecursiveDirectoryWatcher.java

```diff
@@ -0,0 +1,234 @@
+import com.google.common.collect.HashBiMap;
+import com.sk89q.worldedit.WorldEdit;
+import com.sk89q.worldedit.util.io.file.FilenameException;
+import org.apache.logging.log4j.Logger;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.nio.file.ClosedWatchServiceException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.StandardWatchEventKinds;
+import java.nio.file.WatchEvent;
+import java.nio.file.WatchKey;
+import java.nio.file.WatchService;
+import java.util.Optional;
+import java.util.function.Consumer;
+
+/**
+ * Helper class that recursively monitors a directory for changes to files and folders, including creation, deletion, and modification.
+ *
+ * @apiNote File and folder events might be sent multiple times. Users of this class need to employ their own
+ *      deduplication!
+ */
+public class RecursiveDirectoryWatcher implements Closeable {
+
+    /**
+     * Base interface for all change events.
+     */
+    public interface DirEntryChangeEvent {
+        Path path();
+    }
+
+    /**
+     * Event signaling the creation of a new file.
+     */
+    public record FileCreatedEvent(Path path) implements DirEntryChangeEvent {
+    }
+
+    /**
+     * Event signaling the deletion of a file.
+     */
+    public record FileDeletedEvent(Path path) implements DirEntryChangeEvent {
+    }
+
+    /**
+     * Event signaling the creation of a new directory.
+     */
+    public record DirectoryCreatedEvent(Path path) implements DirEntryChangeEvent {
+    }
+
+    /**
+     * Event signaling the deletion of a directory.
+     */
+    public record DirectoryDeletedEvent(Path path) implements DirEntryChangeEvent {
+    }
+
+
+    private static final Logger LOGGER = LogManagerCompat.getLogger();
+
+    private final Path root;
+    private final WatchService watchService;
+    private Thread watchThread;
+    private Consumer<DirEntryChangeEvent> eventConsumer;
+    private BiMap<WatchKey, Path> watchRootMap = HashBiMap.create();
+
+    private RecursiveDirectoryWatcher(Path root, WatchService watchService) {
+        this.root = root;
+        this.watchService = watchService;
+    }
+
+    /**
+     * Create a new recursive directory watcher for the given root folder.
+     * You have to call {@link #start(Consumer)} before the instance starts monitoring.
+     *
+     * @param root Folder to watch for changed files recursively.
+     * @return a new instance that will monitor the given root folder
+     * @throws IOException If creating the watcher failed, e.g. due to root not existing.
+     */
+    public static Optional<RecursiveDirectoryWatcher> create(Path root) throws IOException {
+        try {
+            WatchService watchService = root.getFileSystem().newWatchService();
+            return Optional.of(new RecursiveDirectoryWatcher(root, watchService));
+        } catch (UnsupportedOperationException ignored) {
+            return Optional.empty();
+        }
+    }
```

Comment:
Let's not bother with the `Optional` path here. We're not currently dealing with non-default `Path` providers, and I don't expect to any time soon. Quote from the Javadocs of `newWatchService`:
> This exception is not thrown by FileSystems created by the default provider.

So, let's remove the `Optional` and just throw.

```suggestion
    public static RecursiveDirectoryWatcher create(Path root) throws IOException {
        WatchService watchService;
        try {
            watchService = root.getFileSystem().newWatchService();
        } catch (UnsupportedOperationException e) {
            throw new IllegalArgumentException("Root must support file watching", new RecursiveDirectoryWatcher(root, watchService));
        }
        return new RecursiveDirectoryWatcher(root, watchService);
    }
```